sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,474 @@
+        self._on_persistent_failure = on_persistent_failure_callback
+        self.on_message_returned = on_message_returned_callback
+        self._reconnect_delay = reconnect_delay or self.DEFAULT_RECONNECT_DELAY
+        self._timeout = timeout or self.DEFAULT_TIMEOUT
+        self._connection_attempts = (connection_attempts or
+                                     self.DEFAULT_CONNECTION_ATTEMPTS)
+        self._schemas = {}
+        self._state = self.STATE_IDLE
+        self._url = url
+        self._condition = locks.Condition()
+
+        if self._timeout <= self._reconnect_delay:
+            raise AttributeError('Reconnection must be less than timeout')
+
+        # Automatically start the RabbitMQ connection on creation
+        self._connection = self.connect()
+
+    @gen.coroutine
+    def publish(self, exchange, routing_key, body,
+                app_id=None, correlation_id=None, properties=None,
+                mandatory=False):
+        """Publish a message to RabbitMQ. If the RabbitMQ connection is not
+        established or is blocked, attempt to wait until sending is possible.
+
+        :param str exchange: The exchange to publish the message to.
+        :param str routing_key: The routing key to publish the message with.
+        :param bytes body: The message body to send.
+        :param str app_id: The ID of the app sending the message.
+        :param str correlation_id: The correlation ID associated
+                                   with the message.
+        :param dict properties: An optional dict of additional properties
+                                to append. Will not override mandatory
+                                properties:
+                                app_id, correlation_id, message_id, timestamp
+        :param bool mandatory: Whether to instruct the server to return an
+                               unqueueable message
+                               http://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.publish.mandatory
+
+        """
+        if self.closing:
+            LOGGER.warning('Connection Closing, waiting for reconnection')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif self.closed:
+            LOGGER.warning('Connection Closed, waiting for reconnection')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif self.blocked:
+            LOGGER.warning('Connection Blocked, waiting for unblock')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif not self.ready:
+            LOGGER.warning('Connection not ready, waiting for ready')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+
+        if self.connected:
```

Comment:
Is there any chance that it'll get to this point and not be connected, because then it's fail to publish silently?